src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -3799,6 +3805,159 @@ private bool SelfDescribingEvents
+// it to mean other aspects of tracing such as these EventSources.
+#if FEATURE_PERFTRACING
+            _ = NativeRuntimeEventSource.Log;
+            _ = RuntimeEventSource.Log;
+#endif
+            // System.Diagnostics.MetricsEventSource allows listening to Meters and indirectly
+            // also creates the System.Runtime Meter.
+
+            // Functionally we could preregister NativeRuntimeEventSource and RuntimeEventSource as well, but it may not provide
+            // much benefit. The main benefit for MetricsEventSource is that the app may never use it and it defers
+            // pulling the System.Diagnostics.DiagnosticSource assembly into the process until it is needed.
+            if (AppContext.TryGetSwitch("System.Diagnostics.Metrics.Meter.IsSupported", out bool isSupported) ? isSupported : true)
+            {
+                string name = "System.Diagnostics.Metrics";
+                Guid id = new Guid("20752bc4-c151-50f5-f27b-df92d8af5a61");
+                PreregisterEventProviders(id, name, GetMetricsEventSource);
+            }
+        }
+
+        private static EventSource? GetMetricsEventSource()
+        {
+            Type? metricsEventSourceType = Type.GetType(
+                "System.Diagnostics.Metrics.MetricsEventSource, System.Diagnostics.DiagnosticSource",
+                throwOnError: false);
+
+            if (metricsEventSourceType == null)
+            {
+                return null;
+            }
+            MethodInfo? getInstanceMethod = metricsEventSourceType.GetMethod("GetInstance", BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static, null, Type.EmptyTypes, null);
+            if (getInstanceMethod == null)
+            {
+                return null;
+            }
+            return getInstanceMethod.Invoke(null, null) as EventSource;
+        }
+
+        // Pre-registration creates and registers an EventProvider prior to the EventSource being constructed.
+        // If a tracing session is started using the provider then the EventSource will be constructed on demand.
+        private static unsafe void PreregisterEventProviders(Guid id, string name, Func<EventSource?> eventSourceFactory)
+        {
+            // NOTE: Pre-registration has some minor limitations and variations to normal EventSource behavior:
+            // 1. Instead of delivering OnEventCommand callbacks during the EventSource constructor it may deliver them after
+            //    the constructor instead. This occurs because the EventProvider callback might create the EventSource instance
+            //    in the factory method, then deliver the callback.
+            // 2. EventSource traits aren't supported. Normally EventSource.Initialize() would init provider metadata including
+            //    traits but the SetInformation() call we use below generates metadata from name only. If we want traits support
+            //    in the future it could be added.
+
```

Comment:
Note to myself: Add a comment about ETW's deadlock behavior on Register when called within a callback. That is what requires us to use the provider we already registered rather than setting up a new one inside EventSource.Initialize